### PR TITLE
Merge RecommendedProductDependency

### PR DESCRIPTION
**Pull request by dansanduleac**
Currently, [`CreateManifestTask`](https://github.com/palantir/sls-packaging/blob/develop/gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy) will fail hard (with no recourse) if multiple `RecommendedProductDependency` come from two transitive dependencies and they are not exactly equal (i.e. they have the same minimum, maximum and recommended versions).
This is bad, and causes people to think about workarounds such as #279.
That change might be useful in itself as well (as a last recourse, to get people unblocked when transitive deps have declared unnecessarily strict version ranges in their dependencies).

However, it would be better if we make an attempt to reconcile two conflicting `RecommendedProductDependency` by constructing the most accommodating such dependency entry that respects the min/max versions of the conflicted dependency entries.

Simplified example:
```
{min=1.0.0, max=1.6.x}
{min=1.3.5, max=1.x.x}
-->
{min=1.3.5, max=1.6.x}
```

### CODE REVIEW

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: not sure if I'm doing the right thing here, this is a weird groovy class...

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: nonNull earlier?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy]: pls no

Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy]: yep sorry - it was intellij :)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (ferozco) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: why do we need a class hierarchy?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (ferozco) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: please leave a comment

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: explained in docstring

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: I know this is a cool way to achieve pattern matching-style stuff, but it's pretty confusing to read.  Please could you just use the normal MaximumVersionVisitor thing instead so that it's clear there's no magic here?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: I think the super functional approach here actually obscures readability... could we do this imperatively?

```java
    static MaximumVersion valueOf(String version) {
        Optional<OrderableSlsVersion> maybeOrderable = OrderableSlsVersion.safeValueOf(version);
        if (maybeOrderable.isPresent()) {
            return new VersionMaximumVersion(maybeOrderable.get());
        }

        Optional<SlsVersionMatcher> maybeMatcher = SlsVersionMatcher.safeValueOf(version);
        if (maybeMatcher.isPresent()) {
            return new MatcherMaximumVersion(maybeMatcher.get());
        }

        throw new SafeIllegalArgumentException(
                "Couldn't parse version as an OrderableSlsVersion or an SlsVersionMatcher", 
                UnsafeArg.of("version", version));
    }
```

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MatcherMaximumVersion.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MatcherMaximumVersion.java]: Possibly worth making these two variants inner classes of MaximumVersion?  I think the javadoc should actually just give an example rather than repeating classnames, e.g.:

> Matchers may contain wildcards, e.g. 1.X.X

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java]: This seems like a really nice pure function for dealing with version ranges.  Can we just put it in sls-versions now that it's open source?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependency.groovy
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependency.groovy:64]: 🤦‍♂️ nice

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: I'm not sure how useful the safe-logging preconditions here are, as we're just going to use gradle's SLF4J implementation... I think this means we need `{}` tokens in the message (or maybe `%s` I can never remember)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: kinda funky to use Stream.of here but Collections.max(Arrays.asList()) above... pick one and run with it?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: mega +1 to encapsulating all this logic in nice clean comparators.  Possibly just inline the variable as the class is already nicely named?

File: versions.props
Comment (iamdanfox) [inline versions.props:1]: btw is the `b-duplicate-different-versions-1.0.jar` necessary?

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy
Comment (iamdanfox) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy:83]: woah. Is this really how we construct RecommendedProductDependencies elsewhere??

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: I don't think this is magic, it's pretty clear what this does coming from a functional programming background. (if the logic were confusing though, that's another issue) 
However it's not a lot more code to write visitors so I can do that.

Comment (iamdanfox) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java]: This may look perfectly readable to you, but I think we need to try to make it super super obvious for any future maintainers.  Given that all the conjure union types just use FooVisitor, I think the consistency will be helpful to readers.

Aiming to use the simplest, most boring solution that works here!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java]: yep, good idea

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: This stuff is not passed to a logger, the Safe*Exception classes generate their own message out of the args

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: it sucks that I have to handle the Optional when using `Stream.max`, when i know there will always be a result here

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy
Comment (dansanduleac) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy:83]: To construct, you can just do `RecommendedProductDependency(an inline map)`.
Here I want to construct taking into account some default values, without having to merge maps manually.

In the gradle plugin, the extension gives you a [configuration point that takes a `Closure`](https://github.com/palantir/sls-packaging/blob/85196ddbf8f10c9fb35f77c985602d3ae715415e/gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesExtension.groovy#L25)

Comment (dansanduleac) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy:83]: This is just so I can clone things so I don't have to re-type the same fields that never change... I didn't know a better way to do that in groovy

File: versions.props
Comment (dansanduleac) [reply versions.props:1]: That is used, I just changed one of the versions in there to make them not compatible, such that the merger would kick in

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java]: https://github.com/palantir/sls-version-java/pull/3

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (iamdanfox) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java]: Let's switch to the guava ones for non-server usage

https://github.com/palantir/safe-logging/blob/develop/preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIllegalStateException.java

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy
Comment (iamdanfox) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy]: Mention non-overlapping ranges

**Review by iamdanfox — APPROVED** [review]
Approved pending the exceptions & using the new OSS sls-versions jar

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java
Comment (uschi2000) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java:40]: does this need to be `Comparable<>`?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java
Comment (uschi2000) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersionComparator.java:32]: this seems to violate the `Comparable` interface when o1==o2?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (uschi2000) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java:29]: why not just `merge`?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (uschi2000) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java:47]: "Failed to determine which of dep1/dep2 is larger. This is a bug."

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java
Comment (uschi2000) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java:79]: nit: this comment is redundant since the next line literally says `result.isValid();` . if anything, maybe consider renaming `isValid` to something like `verifyIsValid`. (does it throw?)

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/MaximumVersionComparatorTest.groovy
Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/MaximumVersionComparatorTest.groovy:30]: I don't think the Comparator interface makes sense here. `2.x.x` is neither larger nor smaller than `2.6.x`: it is simultaneously smaller, equal, and larger.

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy
Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy:56]: why is `2.5.0` not a valid merged version here?

**Pull request merged**
